go/src/koding/kites/kloud/provider/google/stack.go

```diff
@@ -0,0 +1,254 @@
+	}
+
+	t, err := newBootstrapTemplate(cfg)
+	if err != nil {
+		return nil, err
+	}
+
+	cred := c.Credential.(*Cred)
+	t.Key = cred.Project + "-" + s.BootstrapArg().GroupName + "-" + c.Identifier
+	return []*stack.Template{t}, nil
+}
+
+func (s *Stack) ApplyTemplate(c *stack.Credential) (*stack.Template, error) {
+	t := s.Builder.Template
+
+	var resource struct {
+		GCInstance map[string]map[string]interface{} `hcl:"google_compute_instance"`
+	}
+
+	if err := t.DecodeResource(&resource); err != nil {
+		return nil, err
+	}
+
+	if len(resource.GCInstance) == 0 {
+		return nil, fmt.Errorf("there are no Google compute instances defined")
+	}
+
+	for resourceName, instance := range resource.GCInstance {
+		// Set default machine type if user didn't define it herself.
+		if mt, ok := instance["machine_type"]; !ok {
+			instance["machine_type"] = defaultMachineType
+		} else {
+			if mtstr, ok := mt.(string); ok && mtstr == "" {
+				instance["machine_type"] = defaultMachineType
+			}
+		}
+
+		// Set default zone if user didn't define it herself.
+		const lookupZone = "${lookup(var.zones, var.google_region)}"
+		if z, ok := instance["zone"]; !ok {
+			instance["zone"] = lookupZone
+		} else {
+			if zstr, ok := z.(string); ok && zstr == "" {
+				instance["zone"] = lookupZone
+			}
+		}
+
+		// Set default image for disk if user didn't define it herself.
+		if _, ok := instance["disk"]; !ok {
+			instance["disk"] = struct {
+				Image string `json:"image" bson:"image" hcl:"image"`
+			}{
+				Image: defualtMachineImage,
+			}
+		}
+
+		// Set default network interface if user didn't define it herself.
+		if _, ok := instance["network_interface"]; !ok {
+			instance["network_interface"] = struct {
+				Network      string   `json:"network" bson:"network" hcl:"network"`
+				AccessConfig struct{} `json:"access_config" bson:"access_config" hcl:"access_config"`
+			}{
+				Network:      "default",
+				AccessConfig: struct{}{},
+			}
+		}
+
+		// means there will be several instances, we need to create a userdata
+		// with count interpolation, because each machine must have an unique
+		// kite id.
+		count := 1
+		if n, ok := instance["count"].(int); ok && n > 1 {
+			count = n
+		}
+
+		labels := []string{resourceName}
+		if count > 1 {
+			labels = labels[:0]
+			for i := 0; i < count; i++ {
+				labels = append(labels, fmt.Sprintf("%s.%d", resourceName, i))
+			}
+		}
+
+		kiteKeyName := fmt.Sprintf("kitekeys_%s", resourceName)
+
+		// Cloud-Init can be injected only via "user-data" field defined in
+		// root "metadata" object.
+		metadata := interfaceToMap(instance["metadata"])
+		s.Builder.InterpolateField(metadata, resourceName, "user-data")
+
+		// this part will be the same for all machines
+		userCfg := &userdata.CloudInitConfig{
+			Username: s.Req.Username,
+			Groups:   []string{"sudo"},
+			Hostname: s.Req.Username, // no typo here. hostname = username
+			KiteKey:  fmt.Sprintf("${lookup(var.%s, count.index)}", kiteKeyName),
```

Comment:
after fixing https://github.com/koding/koding/pull/9185/files#r82092103 the kitekey can be set directly here